src/main/java/net/dries007/tfc/common/blockentities/AnvilBlockEntity.java

```diff
@@ -0,0 +1,350 @@
+import net.dries007.tfc.common.capabilities.forge.Forging;
+import net.dries007.tfc.common.capabilities.forge.ForgingBonus;
+import net.dries007.tfc.common.capabilities.forge.ForgingCapability;
+import net.dries007.tfc.common.capabilities.heat.HeatCapability;
+import net.dries007.tfc.common.capabilities.heat.IHeat;
+import net.dries007.tfc.common.container.AnvilContainer;
+import net.dries007.tfc.common.container.AnvilPlanContainer;
+import net.dries007.tfc.common.container.ISlotCallback;
+import net.dries007.tfc.common.recipes.AnvilRecipe;
+import net.dries007.tfc.common.recipes.TFCRecipeTypes;
+import net.dries007.tfc.common.recipes.WeldingRecipe;
+import net.dries007.tfc.util.Helpers;
+import org.jetbrains.annotations.Nullable;
+
+public class AnvilBlockEntity extends InventoryBlockEntity<AnvilBlockEntity.AnvilInventory> implements ISlotCallback
+{
+    public static final int SLOT_INPUT_MAIN = 0;
+    public static final int SLOT_INPUT_SECOND = 1;
+    public static final int SLOT_HAMMER = 2;
+    public static final int SLOT_CATALYST = 3;
+
+    public static final int[] SLOTS_BY_HAND_EXTRACT = new int[] {SLOT_INPUT_MAIN, SLOT_INPUT_SECOND};
+    public static final int[] SLOTS_BY_HAND_INSERT = new int[] {SLOT_INPUT_MAIN, SLOT_INPUT_SECOND, SLOT_CATALYST};
+
+    private static final Component NAME = new TranslatableComponent("tfc.block_entity.anvil");
+
+    public AnvilBlockEntity(BlockPos pos, BlockState state)
+    {
+        super(TFCBlockEntities.ANVIL.get(), pos, state, AnvilInventory::new, NAME);
+    }
+
+    @Nullable
+    public Forging getMainInputForging()
+    {
+        return inventory.getStackInSlot(AnvilBlockEntity.SLOT_INPUT_MAIN)
+            .getCapability(ForgingCapability.CAPABILITY)
+            .resolve()
+            .orElse(null);
+    }
+
+    /**
+     * @return the provider for opening the anvil plan screen
+     */
+    public MenuProvider planProvider()
+    {
+        return new SimpleMenuProvider(this::createPlanContainer, getDisplayName());
+    }
+
+    /**
+     * @return the provider for opening the main anvil working screen
+     */
+    public MenuProvider anvilProvider()
+    {
+        return this;
+    }
+
+    @Nullable
+    @Override
+    public AbstractContainerMenu createMenu(int containerId, Inventory inventory, Player player)
+    {
+        return AnvilContainer.create(this, player.getInventory(), containerId);
+    }
+
+    @Nullable
+    public AbstractContainerMenu createPlanContainer(int containerId, Inventory inventory, Player player)
+    {
+        return AnvilPlanContainer.create(this, player.getInventory(), containerId);
+    }
+
+    @Override
+    public boolean isItemValid(int slot, ItemStack stack)
+    {
+        return switch (slot)
+            {
+                case SLOT_INPUT_MAIN, SLOT_INPUT_SECOND -> true;
+                case SLOT_HAMMER -> Helpers.isItem(stack, TFCTags.Items.HAMMERS);
+                case SLOT_CATALYST -> Helpers.isItem(stack, TFCTags.Items.FLUX);
```

Comment:
so I assume we're not gonna have a special catalyst relation in the recipe?

Reply:
I wasn't going to, no, as it overcomplicates the slot eligibility, which affects stuff like shift clicking inputs. If it becomes a requested feature we can look into it, but there are notable downsides.